### ISSUE TITLE
unnecessary use of `to_string` for a type with custom `ToString` trait

### ISSUE THREAD

**Issue by xphoniex** [rust-lang/rust-clippy#9317]
### Summary

I'm programmatically commiting a patch to a git repo and then writing the hash (a [`git2::Oid`](https://docs.rs/git2/latest/git2/struct.Oid.html)) in a file:

```
let commit = util::commit(...);
fs::write(path, commit);
```

if I use above code, I'll get bytes written to the file whereas I wanted the string of hash, so naturally I change it to `fs::write(path, commit.to_string())`, however that's throwing a lint error:

```
warning: unnecessary use of `to_string`
   --> ...
    |
810 |             commit.to_string(),
    |             ^^^^^^^^^^^^^^^^^^ help: use: `commit`
    |
    = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
```

solution is to use `format!(...)` instead, but that's pretty ugly.

### Lint Name

unnecessary_to_owned

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

### Version

```text
rustc 1.61.0 (fe5b13d68 2022-05-18)
binary: rustc
commit-hash: fe5b13d681f25ee6474be29d748c65adcd91f69e
commit-date: 2022-05-18
host: x86_64-unknown-linux-gnu
release: 1.61.0
LLVM version: 14.0.0
```

**Comment by Jarcho** [comment]
Note to anyone working on this. `to_string` should only be linted when the type is string-like. e.g. implements `Deref<Target=str/String>` or `AsRef<str/String>`.

**Comment by DesmondWillowbrook** [comment]
I wasn't able to replicate your report with rust v1.61.0 locally nor v1.63.0 ([playground for v1.63.0 - Tools - Clippy ](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8dd84082d027d82a1cc197f115b8d469))
Am I perhaps missing any relevant trait implementations of `Oid`?

**Comment by rcgoodfellow** [comment]
Hitting this too. A simple repro is

main.rs
```rust
fn main() {
    let id = uuid::Uuid::new_v4();

    std::fs::write("bad", &id).unwrap();
    std::fs::write("good", id.to_string()).unwrap();
}
```
Cargo.toml
```toml
[package]
name = "poo"
version = "0.1.0"
edition = "2021"

[dependencies]
uuid = { version = "1.0.0", features = [ "serde", "v4" ] }
```
Demo
```shell
$ cargo run ; cat good; echo ""; cat bad; echo "";
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/poo`
e9d1adb8-a9b2-4864-a2d2-d8095cba34a9
�ѭ���Hd���	\�4�
```
```shell
$ cargo --version
cargo 1.63.0 (fd9c4297c 2022-07-01)
```
```shell
$ cargo clippy
warning: unnecessary use of `to_string`
 --> src/main.rs:5:28
  |
5 |     std::fs::write("good", id.to_string()).unwrap();
  |                            ^^^^^^^^^^^^^^ help: use: `id`
  |
  = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned

warning: `poo` (bin "poo") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.40s
```

**Comment by xphoniex** [comment]
@DesmondWillowbrook 

Hey, I'd already done the preliminary work and prepared a fix. You can skip this issue.

**Comment by xphoniex** [comment]
@rcgoodfellow 

Yes, this is the same exact issue. PR https://github.com/rust-lang/rust-clippy/pull/9329 fixes this.
